src/Bicep.Core/Semantics/FunctionResult.cs

```diff
@@ -0,0 +1,19 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Bicep.Core.TypeSystem;
+
+namespace Bicep.Core.Semantics;
+
+public class FunctionResult
+{
+    public TypeSymbol Type { get; }
+    public object? Value { get; }
```

Comment:
Here you're using `Value` to store a `JToken` in the *Generator, so that it can be converted into a piece of Bicep Syntax in the *Evaluator - this requires using `object?` and performing type checks in the Evaluator.

What about using `SyntaxBase?` as the type for `Value`, and instead performing the syntax conversion in the *Generator method? This would have a few benefits:
1. Avoids the lack of type safety with `object`
2. Gives additional ability to raise diagnostics during the JSON -> Syntax conversion (see comment about size restriction)

Reply:
I wanted to decouple part with Generating Type ("running function") and code for emitting the result. If I would include emitting logic into GenerateResult, each time we run type analysis we would unnecessarily generate bicep syntax.

Reply:
That's a point regarding saving cycles.

I'd definitely prefer type safety here - what about using a `Lazy<SyntaxBase?>` or a `Func<SyntaxBase?>` to allow for deferred execution?

Reply:
I'm still not quite sure if we should put SyntaxBase here. The idea behind was to pass a value from type Builder (JToken) to emitting logic. In fact, SyntaxBase is also not a good idea, the best would be to have a type that bridged syntax and type system with emitting logic. This object is to avoid 'executing' function for the second time during emitting phase.
I feel that changing this to lazy<SyntaxBase> would require rewriting quite a lot of Function system.

Reply:
Sounds good, I'm happy to keep as-is for now. If we get around to working on an IR, then this would be a good place to come and simplify some of the syntax tree manipulation logic in the namespace classes.